File: ChatGPT3_5/ChainResponsibility/round2/Code.py
class Handler:
    def __init__(self):
        self.next_handler = None

    def set_next(self, handler):
        self.next_handler = handler
        return handler

    def handle(self, request):
        if self.next_handler:
            return self.next_handler.handle(request)
        return "Request could not be processed by any handler."

class LocationHandler(Handler):
    def handle(self, request):
        if request.get('location') == 'valid_location':
            print("LocationHandler: Package can be delivered to this location.")
            return "Location verified"
        print("LocationHandler: Passing to the next handler...")
        return super().handle(request)

class PaymentHandler(Handler):
    def handle(self, request):
        if request.get('payment') == 'paid':
            print("PaymentHandler: Payment has been processed successfully.")
            return "Payment processed"
        print("PaymentHandler: Passing to the next handler...")
        return super().handle(request)

class DeliveryHandler(Handler):
    def handle(self, request):
        if request.get('delivery_ready') == True:
            print("DeliveryHandler: Package is being delivered.")
            return "Package delivered"
        print("DeliveryHandler: Passing to the next handler...")
        return super().handle(request)

# Example usage
if __name__ == "__main__":
    location_handler = LocationHandler()
    payment_handler = PaymentHandler()
    delivery_handler = DeliveryHandler()

    location_handler.set_next(payment_handler).set_next(delivery_handler)

    # Valid request
    request_valid = {'location': 'valid_location', 'payment': 'paid', 'delivery_ready': True}
    result = location_handler.handle(request_valid)
    print(f"Result: {result}")  # Output: "Location verified"

    # Invalid location request
    request_invalid = {'location': 'invalid_location', 'payment': 'unpaid', 'delivery_ready': False}
    result = location_handler.handle(request_invalid)
    print(f"Result: {result}")  # Output: "Request could not be processed by any handler."

File: Java_Gemini_Flash/Chain_Responsibility/Round2/Code.java
package Chain_of_responsibility.round2;

interface Handler {
    void handleRequest(PackageRequest request);
}

class PackageRequest {
    private String location;
    private double paymentAmount;

    public PackageRequest(String location, double paymentAmount) {
        this.location = location;
        this.paymentAmount = paymentAmount;
    }

    public String getLocation() {
        return location;
    }

    public double getPaymentAmount() {
        return paymentAmount;
    }
}

class LocationHandler implements Handler {
    private Handler nextHandler;

    public LocationHandler(Handler nextHandler) {
        this.nextHandler = nextHandler;
    }

    @Override
    public void handleRequest(PackageRequest request) {
        if (isValidLocation(request.getLocation())) {
            System.out.println("Location validated.");
        } else {
            System.out.println("Invalid location. Passing to next handler.");
            nextHandler.handleRequest(request);
        }
    }

    private boolean isValidLocation(String location) {
        // Check if the location is valid, for example:
        return !location.isEmpty();
    }
}

class PaymentHandler implements Handler {
    private Handler nextHandler;

    public PaymentHandler(Handler nextHandler) {
        this.nextHandler = nextHandler;
    }

    @Override
    public void handleRequest(PackageRequest request) {
        if (isPaymentProcessed(request.getPaymentAmount())) {
            System.out.println("Payment processed.");
        } else {
            System.out.println("Payment not processed. Passing to next handler.");
            nextHandler.handleRequest(request);
        }
    }

    private boolean isPaymentProcessed(double paymentAmount) {
        // Check if the payment is processed, for example:
        return paymentAmount > 0;
    }
}

class DeliveryHandler implements Handler {

    @Override
    public void handleRequest(PackageRequest request) {
        System.out.println("Package delivered.");
    }
}
